src/cargo/core/workspace.rs

```diff
@@ -1841,3 +1845,69 @@ impl InheritableFields {
             })
     }
 }
+
+fn parse_manifest(manifest_path: &Path, config: &Config) -> CargoResult<EitherManifest> {
+    let key = manifest_path.parent().unwrap();
+    let source_id = SourceId::for_path(key)?;
+    let (manifest, _nested_paths) = read_manifest(manifest_path, source_id, config)?;
+    Ok(manifest)
+}
+
+pub fn find_workspace_root(manifest_path: &Path, config: &Config) -> CargoResult<Option<PathBuf>> {
```

Comment:
Can we share some of the core logic between `Workspace::find_root` and this?

There is very specific policy common between these and no indication of the coupling

Reply:
The biggest problem with sharing is that `Workspace::find_root` takes a `&mut self` and uses `*self.packages.load()`. One of the ways to get around this would be to delay parsing of a `Cargo.toml` into a manifest and only resolve its `workspace_config` and then fully resolve it after we have found a root workspace. This design was similar to what was done in my last attempt and [wasn't seen as a great design](https://github.com/rust-lang/cargo/pull/9684/#discussion_r728456955) but it also had an `IntermediateManifest`. If there was a way to access `.load()` or early return if it's not a workspace root this could work but it seems significantly harder and could lead to passing things around that we don't need to. 

One idea would be to make a function that only checks a `Cargo.toml`'s `workspace_config`. This would mean that you would read a manifest from the disk more than once unless you kept all partially parsed manifests somewhere to be fully done at a later time. I don't like this way but it is something to consider. 

Unless we come up with a better idea in the meantime I will look into what amount we can share between the two places and try to implement it. If you have any suggestions on how we could change this to work better I'm all ears.

Reply:
At minimum, I'd like the linkage between the two functions clear.

We could make an iterator of possible parents
- handle `paths::ancestors(manifest_path, None).skip(2)`
- handle `target/package`
- handle `config.home() == path`

This just leaves the parsing and excluding logic to be different

Like with my other comments, I'd recommend extracting the iterator in a commit below adding the inheritable fields that then reuses that iterator.

Reply:
(note, you might not need to create a struct to `impl Iterator` but might be able to create a function that `-> impl Iterator`)

Reply:
I'm not exactly sure what you mean. Does this align with what you are trying to say? 
```rust
pub fn temp<'a>(config: &'a Config, manifest_path: &'a Path) -> impl Iterator<Item = PathBuf> + 'a {
    paths::ancestors(manifest_path, None)
        .skip(2)
        .into_iter()
        .filter(|path| !path.ends_with("target/package") || &config.home() != path )
        .map(|path| path.join("Cargo.toml"))
        .filter(|ances_manifest_path| ances_manifest_path.exists())

}
```

Reply:
Roughly.  By using `filter`, you are treating the two `break` cases as `continue`s.  `take_while` for one will fix it but I haven't thought through what would fix the other or if that will force an explicit iterator type to be needed.
```rust
pub fn temp<'a>(config: &'a Config, manifest_path: &'a Path) -> impl Iterator<Item = PathBuf> + 'a {
    paths::ancestors(manifest_path, None)
        .skip(2)
        .into_iter()
        .take_while(|path| !path.ends_with("target/package"))  // now is a break
        .filter(|path| &config.home() != path )  // this is a continue, should be a break *after* yielding the element
        .map(|path| path.join("Cargo.toml"))
        .filter(|ances_manifest_path| ances_manifest_path.exists())
}
```

Reply:
Does that fix it?
```rust
pub fn temp<'a>(config: &'a Config, manifest_path: &'a Path) -> impl Iterator<Item = PathBuf> + 'a {
    paths::ancestors(manifest_path, None)
        .skip(2)
        .into_iter()
        .take_while(|path| !path.ends_with("target/package"))  // now is a break
        .map(|path| path.join("Cargo.toml"))
        .filter(|ances_manifest_path| ances_manifest_path.exists())
        .take_while(|pathbuf| &config.home() != &pathbuf.parent().unwrap())
}
```

Reply:
I don't think so but you can always verify it

Reply:
What I put won't work, we need to be able to grab one past any of the `_while()` methods. I didn't find anything that would work for that. Do think we should make an explicit iterator type then? 

There also is a problem with filtering for 
```rust
.filter(|ances_manifest_path| ances_manifest_path.exists())
```
since you still want to break if it equals `CARGO_HOME`

Reply:
If you also didn't find anything, probably going to require an explicit iterator

Reply:
Either that or some `Fn -> _` that we call in the for loop. I'm gonna go with the explicit iterator first unless you think otherwise.